llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -525,26 +650,149 @@ void propagateAspectsThroughCG(Function *F, CallGraphTy &CG,
   AspectsSetTy LocalAspects;
   for (Function *Callee : It->second) {
     if (Visited.insert(Callee).second)
-      propagateAspectsThroughCG(Callee, CG, AspectsMap, Visited);
+      propagateAspectsThroughCG(Callee, CG, AspectsMap,
+                                FunctionsCalledUnconditionally, Visited);
 
     const auto &CalleeAspects = AspectsMap[Callee];
     LocalAspects.insert(CalleeAspects.begin(), CalleeAspects.end());
   }
+  if (FunctionsCalledUnconditionally.contains(F))
+    AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+}
+
+using PathsContainingConditionalCallersTy =
+    std::vector<std::vector<Function *>>;
+
+void identifyPathsContainingConditionalCallers(
+    Function *F, const CallGraphTy &CG, std::vector<Function *> &CurrentPath,
+    PathsContainingConditionalCallersTy &Paths) {
+  const auto It = CG.find(F);
+  if (It == CG.end()) {
+    CurrentPath.push_back(F);
+    for (auto Func : CurrentPath)
+      if (Func->hasFnAttribute("sycl-call-if-on-device-conditionally")) {
+        Paths.push_back(CurrentPath);
+        break;
+      }
+    CurrentPath.pop_back();
+    return;
+  }
 
-  AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+  CurrentPath.push_back(F);
+
+  for (Function *Callee : It->second) {
+    identifyPathsContainingConditionalCallers(Callee, CG, CurrentPath, Paths);
+  }
+
+  CurrentPath.pop_back();
+}
+
+void propagateConditionExpressionsAndConditionalAspectsThroughCG(
+    const PathsContainingConditionalCallersTy &Paths,
+    FuncToConditionExprsAndConditionalAspectsTy &ConditionsAndAspectsMap,
+    FunctionToConditionExprsMapTy &ConditionsMap,
+    FunctionToAspectsMapTy &AspectsMap) {
+  auto Contains =
+      [](const ConditionExprsAndConditionalAspectsTy &ConditionsAndAspectsVec,
+         const std::pair<ConditionsSetTy, AspectsSetTy> &Search) {
+        return std::any_of(
+            ConditionsAndAspectsVec.begin(), ConditionsAndAspectsVec.end(),
```

Comment:
[`llvm::any_of`](https://llvm.org/doxygen/namespacellvm.html#a61d13d6824ec46c31260a4fd0997eda0) contains helpers which accept ranges, meaning that  you don't need to call `.begin()/.end()` explicitly, but instead can pass a single argument `ConditionsAndAspectsVec`.

You can find more helpers like this in `stl_extras` header